src/Shared/Buffers.MemoryPool/UnmanagedBufferAllocator.cs

```diff
@@ -0,0 +1,156 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+using System.Text;
+
+/// <summary>
+/// Allocator that manages blocks of unmanaged memory.
+/// </summary>
+internal unsafe struct UnmanagedBufferAllocator : IDisposable
+{
+    private readonly int _blockSize;
+    private int _currentBlockCount;
+    private void** _currentAlloc;
+    private byte* _currentBlock;
+
+    /// <summary>
+    /// The default block size for the allocator.
+    /// </summary>
+    /// <remarks>
+    /// This size assumes a common page size and provides an accomdation
+    /// for the pointer chain used to managed the allocations.
+    /// </remarks>
+    public static int DefaultBlockSize => 4096 - sizeof(void*);
+
+    /// <summary>
+    /// Instantiate an <see cref="UnmanagedBufferAllocator"/> instance.
+    /// </summary>
+    /// <param name="blockSize">The unmanaged memory block size in bytes.</param>
+    public UnmanagedBufferAllocator(int blockSize = 0)
+    {
+        Debug.Assert(_blockSize >= 0);
+        _blockSize = blockSize == 0 ? DefaultBlockSize : blockSize;
+        _currentBlockCount = -1;
+        _currentAlloc = null;
+        _currentBlock = null;
+    }
+
+    /// <summary>
+    /// Allocate the requested amount of space from the allocator.
+    /// </summary>
+    /// <typeparam name="T">The type requested</typeparam>
+    /// <param name="count">The count in <typeparamref name="T"/> units</param>
+    /// <returns>A pointer to the reserved memory.</returns>
+    public T* Alloc<T>(int count) where T : unmanaged
+    {
+        int toAlloc = count *sizeof(T);
```

Comment:
```suggestion
        int toAlloc = checked(count * sizeof(T));
```
to avoid silent integer overflow that will lead to a buffer overrun.